category/functor.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2017 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+
+Standard identity and composition functors
+-/
+import tactic.ext
+import category.basic
+
+universe variables u v w u' v' w'
+
+section functor
+
+variables {F : Type u → Type v}
+variables {α β γ : Type u}
+variables [functor F] [is_lawful_functor F]
+
+lemma functor.map_id : functor.map id = (id : F α → F α) :=
+by apply funext; apply id_map
+
+lemma functor.map_comp_map (f : α → β) (g : β → γ) :
+  (functor.map g ∘ functor.map f : F α → F γ) = functor.map (g ∘ f) :=
+by apply funext; intro; rw comp_map
+
+@[norm]
+lemma functor.map_map (f : α → β) (g : β → γ) (x : F α) :
+  g <$> f <$> x = (g ∘ f) <$> x :=
+by rw ← comp_map
+
+end functor
+
+def id.mk {α : Sort u} : α → id α := id
+
+namespace functor
+
+structure comp (f : Type u → Type u') (g : Type v → Type u) (α : Type v) : Type u' :=
```

Comment:
can you again write down a problem when using `function.comp`?

Reply:
It's because of type class inference. Lean is pretty aggressive in introducing lambda abstractions / `id` so a situation where `functor f` would do can lead Lean to try `functor (f o id)` and run out of steam trying to make that work.